quinn/src/streams.rs

```diff
@@ -399,36 +370,25 @@ where
         &mut self,
         cx: &mut Context,
     ) -> Poll<Result<Option<(Bytes, u64)>, ReadError>> {
-        self.any_data_read = true;
-        use proto::ReadError::*;
-        let mut conn = self.conn.lock().unwrap();
-        if self.is_0rtt {
-            conn.check_0rtt().map_err(|()| ReadError::ZeroRttRejected)?;
-        }
-        match conn.inner.read_unordered(self.stream) {
-            Ok(Some((bytes, offset))) => {
-                // Flow control credit may have been issued
-                conn.wake();
-                Poll::Ready(Ok(Some((bytes, offset))))
-            }
-            Ok(None) => {
-                self.all_data_read = true;
-                Poll::Ready(Ok(None))
-            }
-            Err(Blocked) => {
-                if let Some(ref x) = conn.error {
-                    return Poll::Ready(Err(ReadError::ConnectionClosed(x.clone())));
-                }
-                conn.blocked_readers.insert(self.stream, cx.waker().clone());
-                Poll::Pending
-            }
-            Err(Reset(error_code)) => {
-                self.all_data_read = true;
-                Poll::Ready(Err(ReadError::Reset(error_code)))
-            }
-            Err(UnknownStream) => Poll::Ready(Err(ReadError::UnknownStream)),
-            Err(IllegalOrderedRead) => Poll::Ready(Err(ReadError::IllegalOrderedRead)),
-        }
+        self.poll_read_generic(cx, |conn, stream| conn.inner.read_unordered(stream))
+    }
+
+    /// Read the next segment of data from next valid offset in the stream.
+    ///
+    /// Yields a segment of data consecutive to the last ordered read offset, or `None`
+    /// if the stream was finished. Segments are received in order.
+    ///
+    /// Beware, the returned chunks bounds won't always match those written by the peer.
+    ///
+    /// Chunk reads have reduced overhead and higher throughput, and should therefore be
+    /// preferred when applicable.
```

Comment:
Tweaked wording a bit:
```suggestion
    /// Yields a segment of data beginning immediately after the last data yielded by
    /// `read` or `read_chunk`, or `None` if the stream was finished.
    ///
    /// Slightly more efficient than `read` due to not copying. Chunks boundaries
    /// do not correspond to peer writes, and hence cannot be used as framing.
```